Use an external scanner to recognise more kinds of syntax

leonerd opened the pull request:
Adds recognition for many kinds of strings; `''`, `q()`, `""` and `qq()`; with balanced delimiters and counting within them.

Recognises `qw(...)` lists.

Also applies special handling of semicolon at the end of a block so the final `;` is not required there.

rabbiveesh commented on grammar.js:
```diff
@@ -194,7 +208,7 @@ module.exports = grammar({
     slice_expression: $ => choice(
       seq('(', optional(field('list', $._expr)), ')', '[', $._expr, ']'),
-      // TODO: QWLIST
+      seq(field('list', $.quoted_word_list), '[', $._expr, ']'),
```
i have learned a new syntax, lol

rabbiveesh commented on grammar.js:
```diff
@@ -268,6 +282,10 @@ module.exports = grammar({
       // legacy
       $.primitive,
+
+      // new scanner temporary testing
+      $.string_literal,
+      $.interpolated_string_literal,
```
are these meant to still be here? not sure the timespan on the "temporary"

leonerd replied:
Oh, yes. They're part of what `toke.c` calls a `THING` and I'm not sure I want to call it that here. But I'll find a better name / comment for this for now.

leonerd replied:
Done

rabbiveesh commented on src/scanner.c:
```diff
@@ -0,0 +1,376 @@
+  int delim_count;
+};
+
+#define TOKEN(type) \
+  do {                            \
+    DEBUG("token(%s)\n", #type);  \
+    lexer->result_symbol = type;  \
+    return true;                  \
+  } while(0)
```
i feel like i'm back in perl land with all the macros!

THX for the PR!

leonerd replied:
I don't know *what* you could mean ;)
https://github.com/Perl/perl5/blob/blead/toke.c#L231

rabbiveesh commented on src/scanner.c:
```diff
@@ -0,0 +1,376 @@
+#include <wctype.h>
+
+enum TokenType {
+  /* ident-alikes */
```
what exactly are ident-alikes?

rabbiveesh replied:
actually, i got it

rabbiveesh commented on src/scanner.c:
```diff
@@ -0,0 +1,376 @@
+
+      TOKEN(PERLY_SEMICOLON);
+    }
+    if(!c || c == '}') {
+      DEBUG("Fake PERLY_SEMICOLON at end-of-scope\n", 0);
+      // no advance
+
+      TOKEN(PERLY_SEMICOLON);
+    }
```
are you able to check for EOF also?

leonerd replied:
Oh; my intention with `!c` was to do exactly that, but maybe it could be more explicit

leonerd replied:
Done

rabbiveesh commented on src/scanner.c:
```diff
@@ -0,0 +1,376 @@
+    bool valid = false;
+
+    int c;
+    while((c = lexer->lookahead)) {
+      if(iswspace(c))
+        break;
+      if(state->delim_open && c == state->delim_open)
+        state->delim_count++;
+      else if(c == state->delim_close) {
+        if(state->delim_count)
+          state->delim_count--;
+        else
+          break;
+      }
+
+      lexer->advance(lexer, false);
+      valid = true;
+    }
```
this looks like it's missing support for escaping the end token

leonerd replied:
Yes; I see you added that but actually I think it should be done a different way. Escaped delimiters are not the same thing as escape sequences. Also regular escape sequences like `\n` don't have any meaning in QWLIST.

leonerd replied:
I've added some extra logic and some more tests around that case

rabbiveesh reviewed (changes requested):
a few small things

leonerd commented:
Fixes applied; PTAL

We may want to squash a few of these commits down - e.g. All three of the `qw()` ones (e30ff4d, 3a3ac2c, aee6403)

rabbiveesh commented on src/scanner.c:
```diff
@@ -326,12 +360,27 @@ bool tree_sitter_perl_external_scanner_scan(
         break;
-      if(state->delim_open && c == state->delim_open)
+
+      if(c == '\\') {
+        /* Most escapes don't count inside a qw() list, but escaped delimiters
+         * still do. That is to say, a '\n' is taken as literal, but '\('
+         * counts as just '('. We need to handle this carefully
+         */
+        lexer->mark_end(lexer);
+        lexer->advance(lexer, false);
+
+        c = lexer->lookahead;
+        if(c == state->delim_open || c == state->delim_close)
+          TOKEN(TOKEN_QW_LIST_CONTENT);
+      }
+      else if(state->delim_open && c == state->delim_open)
```
added some failing tests for missing behavior here; see most recent commit

leonerd replied:
Ahyes, it wasn't correctly handling a QW_LIST_CONTENT fragment that begins with `\`. I've now fixed that and that test passes.

Your other test, of an `\n` inside a `''` string, looked to be expecting the wrong thing. A `\n` inside such a string is an escape sequence so it should be recognised as such.

Code/tests updated.

I think we should squash it down before merge though - shall I do that?

rabbiveesh replied:
as far as I understood '\n' does not make a newline. I don't think you'd want that to be highlighted like an escape sequence.

rabbiveesh reviewed (changes requested):
some edge cases left

leonerd commented:
I'd like to get #11 merged before this one, so we can rebase this atop that and add lots of highlight tests for all these string literals

leonerd commented:
OK. Totally rebased, squashed down a bit, and added highlights tests.

Also fixed up the issue of `\n` etc being an escape-sequence inside `q()` strings. That's now no longer happens, but `\\` is still recognised specially. I've added tree and highlight tests for that case too.

I believe this should be good to go now.

rabbiveesh commented on src/scanner.c:
```diff
@@ -0,0 +1,426 @@
+    if(c == '\\')
+      TOKEN(TOKEN_ESCAPE_SEQUENCE);
+
+    if(valid_symbols[TOKEN_Q_STRING_CONTENT]) {
+      // Inside a q() string, only \\ is a valid escape sequence; all else is literal
+      TOKEN(TOKEN_Q_STRING_CONTENT);
+    }
```
this is true for `qw` also

leonerd replied:
Added

rabbiveesh reviewed (changes requested):
i think this is actually gonna be the last change

rabbiveesh reviewed (approved):
okay let's go!
